### DISCUSSION TITLE
Can't find variable when using subcomponent

### DISCUSSION THREAD

**Discussion by joostvanhoof** [alpinejs/alpine#1876] [1. Help]
I was in doubt whether this should go under Livewire or AlpineJS, but I think it's the latter.

My setup is like this:

**In a blade view `example.blade.php` somewhere in the galaxy:**
```html
<div x-data="{
@foreach($posts as $post)
  postActiveId{{ $post->id }}: false,
@endforeach
}"

  <div x-show="postActiveId1"></div> <! –– this works ––>

  <x-dashboard.test /> <! –– including a component here ––>

</div>
```

If I do the same `x-show` in the component `dashboard.test` that was included above:

```html
<div x-show="postActiveId1"></div>
```

the functionality of `x-show` _actually works_, but I do get a `Unhandled Promise Rejection: ReferenceError: Can't find variable: postActiveId1`

The `example.blade.php` is a Livewire component so maybe that has to do with it. I tried adding `wire:key="{{ $post->id }}"` to the div in the `dashboard.test` component but that didn't resolve it.

**Comment by KevinBatdorf** [comment]
What's rendered to the markup inside `x-data`?

**Reply by joostvanhoof** [reply]
Sorry I don't think I understand what you mean?

**Reply by KevinBatdorf** [reply]
When you load your page and look what's in the dom, what do you see? 

![Screen Shot 2021-07-30 at 2 10 02 PM](https://user-images.githubusercontent.com/1478421/127694571-e82ad299-468d-4302-84e2-8fe31381f218.png)

**Reply by joostvanhoof** [reply]
Ah right, got it could've gotten that the first time :-). It renders the variables correctly:

```html
<div x-data="{
  postActiveId1: false,
  postActiveId2: false,
  postActiveId3: false,
}">
```

**Reply by KevinBatdorf** [reply]
Okay, just checking. It's throwing a `Unhandled Promise Rejection:` which means 1. A promise is being rejected for some reason, and 2 that rejection has a callback that is trying to access the component data but cannot. 

Do you have more methods you didn't share or is everything via LIvewire? 

It may be something you need to ask in the Livewire repo as people watching there have more knowledge of the internals.

**Reply by joostvanhoof** [reply]
Yeah it's a little tricky. Given that the first `x-show` works, but the second (within the component) doesn't made me think it was AlpineJS, but maybe indeed it's related to Livewire.

**Reply by joostvanhoof** [reply]
I know that @joshhanley is the man, and we've been discussing some AlpineJS + Livewire behaviour recently. Josh do you maybe have an idea what's happening here? As indicated I tried adding `wire:key` but that didn't resolve the issue. The weird things is that everything does work as expected. It seems like the `x-show` is executed first before `x-data` (because of rendering order?) and therefore it can't find the variables, but not sure this is how AlpineJS works.

**Reply by joostvanhoof** [reply]
Hmm it looks like it's not the component, outside of the component it happens as well. My guess is that because `x-data` is built dynamically it's called with `x-show` before it's done rendering those variables.

In my example I simplified, but to build the `x-show` I use a loop as well:

```html
@foreach($posts as $post)
  <div x-show="postActiveId{{ $post->id }}"></div>
@endforeach
```

if I add a `wire:key` to that div it also doesn't work.

**Reply by KevinBatdorf** [reply]
Can you set it up on the [laravel playground](https://laravelplayground.com/#/gist/286de510bfc0a88e697284e90ed1d7da)?

**Comment by joshhanley** [comment]
@joostvanhoof did you manage to sort this out?

**Reply by joostvanhoof** [reply]
Nope I still haven't figured it out. It looks like I can not include components like `<x-test></x-test>` in the Laravel playground?

**Reply by joshhanley** [reply]
@joostvanhoof hmm I thought you could, but if you can't get the playground working, I'd be happy to have a look at an example repo if that's easier?

**Reply by joostvanhoof** [reply]
@joshhanley yeah would love to debug this. It's work in progress and quite a repo, happy to walk you through it. Would you be up for a 20 min video call to pair on this?

**Reply by joshhanley** [reply]
@joostvanhoof don't have the time at the moment. But if you can put a small test repo together that I could pull down, I'll have a look when I get a chance 🙂

**Reply by joostvanhoof** [reply]
Thanks. I'll see if I can get that done soon, will keep you posted!

**Reply by joshhanley** [reply]
No worries!
